File: scripts/mcp-conformance/client_against_aha.py
"""Conformance test: official MCP SDK client → aha mcp server.

Spawns `aha mcp` and drives it through the official Anthropic MCP Python
SDK (`mcp.ClientSession` + `mcp.client.stdio.stdio_client`). The SDK speaks
the real MCP stdio wire format (newline-delimited JSON), so a successful
run proves aha's server is wire-compliant. Tested handshake, tool listing,
and one round-trip tool call against a corpus populated in setup.

Skips with exit 77 (autotools convention) if AHA_CONFIG isn't set; the
verify script supplies that path after building a fixture corpus.
"""

import asyncio
import json
import os
import sys

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

# Tools aha exposes today. Update this list when readOnlyTools in
# internal/mcp/tools.go changes — the test asserts exact equality.
EXPECTED_TOOLS = [
    "search",
    "read",
    "status",
    "verify",
    "conflicts",
    "corpus_size",
    "doctor",
]


async def main() -> int:
    aha_bin = os.environ.get("AHA_BIN", "/tmp/aha")
    config = os.environ.get("AHA_CONFIG")
    if not config:
        print("AHA_CONFIG unset; skipping", file=sys.stderr)
        return 77

    params = StdioServerParameters(command=aha_bin, args=["mcp", "--config", config])

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            init = await session.initialize()
            assert init.serverInfo.name == "aha", f"serverInfo.name={init.serverInfo.name!r}"
            print(f"initialize OK: protocol={init.protocolVersion} server={init.serverInfo.name} {init.serverInfo.version}")

            listed = await session.list_tools()
            got = sorted(t.name for t in listed.tools)
            want = sorted(EXPECTED_TOOLS)
            assert got == want, f"tools/list mismatch: got={got} want={want}"
            print(f"tools/list OK: {got}")

            # Read-only hint must be advertised. (2025-06-18 spec.)
            for t in listed.tools:
                ann = (t.annotations.model_dump() if t.annotations else {})
                assert ann.get("readOnlyHint") is True, f"{t.name}: readOnlyHint not True ({ann})"
            print(f"tool annotations OK: readOnlyHint=true on all {len(listed.tools)} tools")

            # Round-trip a no-arg tool that always succeeds against an empty corpus.
            res = await session.call_tool("status", {})
            assert res.content, f"call_tool(status) returned no content blocks"
            text = res.content[0].text
            status = json.loads(text)
            for key in ("corpus_dir", "sessions", "entries", "fts_messages"):
                assert key in status, f"status missing {key!r}: {list(status.keys())[:10]}"
            print(f"tools/call status OK: sessions={status.get('sessions')} entries={status.get('entries')}")

            # If the server exposes structuredContent (spec 2025-06-18), it
            # should equal the parsed text payload.
            if getattr(res, "structuredContent", None) is not None:
                assert res.structuredContent == status, "structuredContent != parsed text content"
                print("tools/call structuredContent OK: matches text content")

            # Tool that takes args. The fixture seeded a user message "hello"
            # so the search guarantees at least one hit; otherwise we couldn't
            # tell empty-result handling from broken framing.
            res = await session.call_tool("search", {"query": "hello", "limit": 5})
            assert res.content, "call_tool(search) returned no content"
            hits = json.loads(res.content[0].text)
            assert isinstance(hits, list), f"search result is not a list: {type(hits).__name__}"
            assert len(hits) >= 1, f"search('hello') returned no hits: {hits}"
            assert all("ref_text" in h for h in hits), f"hit missing ref_text: {hits[0]}"
            print(f"tools/call search OK: returned {len(hits)} hits")

            # Empty-result query: serialized payload should be an empty
            # list, not JSON null. Guards against Go's nil-slice → null
            # marshalling regression.
            res = await session.call_tool("search", {"query": "definitelynotinthecorpus", "limit": 5})
            empty = json.loads(res.content[0].text)
            assert empty == [], f"empty search result not []: {empty!r}"
            print(f"tools/call search empty-result OK: payload is []")

            # Strict-validation: unknown arg must be rejected.
            try:
                await session.call_tool("search", {"query": "x", "bogus": 1})
            except Exception as e:
                msg = str(e)
                assert "unexpected argument" in msg or "bogus" in msg, f"unexpected error: {msg!r}"
                print(f"tools/call unknown-arg rejection OK")
            else:
                print("FAIL: unknown-arg search call was accepted", file=sys.stderr)
                return 1

            # Unknown tool must error.
            try:
                await session.call_tool("refresh", {})
            except Exception as e:
                print(f"tools/call unknown-tool rejection OK")
            else:
                print("FAIL: unknown tool call was accepted", file=sys.stderr)
                return 1

    print("aha MCP server is conformant under the official Python SDK client.")
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: scripts/mcp-conformance/client_against_aha.ts
// Conformance test: official @modelcontextprotocol/sdk Client → aha mcp.
//
// Mirrors scripts/mcp-conformance/client_against_aha.py but uses the
// TypeScript SDK as the validating client. A green run proves aha's
// server is interoperable with both Anthropic-maintained SDKs.
//
// Run via `scripts/verify.sh mcp` (it spawns a fresh corpus first); or
// Direct ambient AHA_BIN/AHA_CONFIG values are rejected; run through the
// attested private workspace created by scripts/verify.sh mcp.

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { assertAttestedConformance } from "./attestation.ts";

const EXPECTED_TOOLS = [
  "search",
  "read",
  "incidents",
  "incident_trajectory",
  "overview",
  "status",
  "verify",
  "conflicts",
  "corpus_size",
  "doctor",
];

const { ahaBin, ahaConfig } = assertAttestedConformance();

const transport = new StdioClientTransport({
  command: ahaBin,
  args: ["mcp", "serve", "--config", ahaConfig],
});

const client = new Client({ name: "aha-conformance-ts", version: "0.1.0" });

async function main() {
  await client.connect(transport);
  const init = client.getServerVersion();
  if (!init) throw new Error("no serverInfo after connect");
  if (init.name !== "aha") {
    throw new Error(`serverInfo.name=${JSON.stringify(init.name)}, want aha`);
  }
  console.log(`initialize OK: server=${init.name} ${init.version}`);

  // tools/list
  const listed = await client.listTools();
  const got = listed.tools.map((t) => t.name).sort();
  const want = [...EXPECTED_TOOLS].sort();
  if (JSON.stringify(got) !== JSON.stringify(want)) {
    throw new Error(`tools/list mismatch:\n  got:  ${got.join(",")}\n  want: ${want.join(",")}`);
  }
  console.log(`tools/list OK: ${got.join(", ")}`);

  // readOnlyHint must be advertised per 2025-06-18.
  for (const t of listed.tools) {
    if (!t.annotations || t.annotations.readOnlyHint !== true) {
      throw new Error(`${t.name}: readOnlyHint not true (${JSON.stringify(t.annotations)})`);
    }
  }
  console.log(`tool annotations OK: readOnlyHint=true on all ${listed.tools.length} tools`);

  // tools/call status — object-returning tool. structuredContent should
  // be present and match the text payload.
  const statusRes = await client.callTool({ name: "status" });
  if (!Array.isArray(statusRes.content) || statusRes.content.length === 0) {
    throw new Error("status: no content blocks");
  }
  const block = statusRes.content[0] as { type: string; text: string };
  const status = JSON.parse(block.text);
  for (const key of ["workspace_dir", "sessions", "entries", "fts_messages"]) {
    if (!(key in status)) throw new Error(`status missing ${key}`);
  }
  console.log(`tools/call status OK: sessions=${status.sessions} entries=${status.entries}`);

  if ("structuredContent" in statusRes && statusRes.structuredContent !== undefined) {
    const sc = statusRes.structuredContent;
    if (JSON.stringify(sc) !== JSON.stringify(status)) {
      throw new Error(`structuredContent != parsed text content`);
    }
    console.log("tools/call structuredContent OK: matches text content");
  }

  // tools/call search — list-returning tool. Text payload must be a list.
  const sr = await client.callTool({ name: "search", arguments: { query: "hello", limit: 5 } });
  const hits = JSON.parse((sr.content as { text: string }[])[0].text);
  if (!Array.isArray(hits)) throw new Error(`search not a list: ${typeof hits}`);
  if (hits.length < 1) throw new Error(`search('hello') returned no hits: ${JSON.stringify(hits)}`);
  if (!("ref_text" in hits[0])) throw new Error(`hit missing ref_text: ${JSON.stringify(hits[0])}`);
  console.log(`tools/call search OK: returned ${hits.length} hits`);

  // Empty result must marshal as [], not null.
  const er = await client.callTool({
    name: "search",
    arguments: { query: "definitelynotinthecorpus", limit: 5 },
  });
  const empty = JSON.parse((er.content as { text: string }[])[0].text);
  if (!Array.isArray(empty) || empty.length !== 0) {
    throw new Error(`empty search result not []: ${JSON.stringify(empty)}`);
  }
  console.log("tools/call search empty-result OK: payload is []");

  // Strict argument validation: unknown arg must surface as isError on the
  // result (the spec's reporting mechanism for tool-level failures).
  {
    const res = (await client.callTool({
      name: "search",
      arguments: { query: "x", bogus: 1 } as Record<string, unknown>,
    })) as { isError?: boolean; content?: { text: string }[] };
    if (!res.isError) throw new Error("unknown-arg search call was accepted (no isError)");
    const err = res.content?.[0]?.text ?? "";
    if (!err.includes("unexpected argument") && !err.includes("bogus")) {
      throw new Error(`unexpected error message: ${err}`);
    }
    console.log("tools/call unknown-arg rejection OK");
  }

  // Unknown tool: SDKs may raise OR return isError.
  {
    let detected = false;
    try {
      const res = (await client.callTool({ name: "refresh", arguments: {} })) as { isError?: boolean };
      detected = !!res.isError;
    } catch {
      detected = true;
    }
    if (!detected) throw new Error("unknown tool call was accepted");
    console.log("tools/call unknown-tool rejection OK");
  }

  console.log("aha MCP server is conformant under the official TypeScript SDK client.");
}

main()
  .then(() => transport.close())
  .then(() => process.exit(0))
  .catch((err) => {
    console.error(err);
    transport.close().catch(() => {});
    process.exit(1);
  });
